src/Common/RWLock.h

```diff
@@ -34,6 +34,10 @@ using RWLock = std::shared_ptr<RWLockImpl>;
 /// - SELECT thread 1 locks in the Read mode
 /// - ALTER tries to lock in the Write mode (waits for SELECT thread 1)
 /// - SELECT thread 2 tries to lock in the Read mode (waits for ALTER)
+///
+/// NOTE: it is dangerous to acquire lock with NO_QUERY, because FastPath doesn't
```

Comment:
Why we are using this RWLock?
`std::shared_mutex` may suffice.

Reply:
And why rwlock at all?

Reply:
I think, because of [this](https://github.com/ClickHouse/ClickHouse/blob/e9749b00275f882d04adf28c31e90a638f8b1904/src/Common/RWLock.h#L26-L30). Looks still reasonable.

Reply:
Maybe we can try to get rid of this RWLock by using some flag, that means, that table cannot be dropped at this moment and a condvar, because now the only usage of exclusive lock are `DROP` and `TRUNCATE` queries.

Reply:
But this construction is almost a RWLock :)

Reply:
+

Fairness is no longer needed.